using-timescaledb/continuous-aggregates.md

```diff
@@ -248,6 +248,124 @@ hypertable with a continuous aggregate.
 
 ---
 
+### Advanced Topics [](using-cagg-advanced)
+
+#### Continuous Aggregates for Integer-Based Hypertables [](create-integer)
+
+Usually, continuous aggregates are defined on a time column, but it is
+also possible to create your own custom scheme for handling
+aggregation for tables that are using an integer for the time
+column. This can be useful if you have tables that use other measures
+of time that can be represented as integer values, such as nanosecond
+epochs, minutes since founding date, or whatever is suitable for your
+application.
+
+As an example, suppose that you have a table with CPU and disk usage
+for some devices where time is measured in
+[microfortnights][fff-system]. Since we are using an integer-valued
+column as time, we need to provide the chunk time interval when
+creating the hypertable. In this case, we let each chunk consist of a
+millifortnight.
+
+```sql
+CREATE TABLE devices(
+  time BIGINT,        -- Timestamp in microfortnights
+  cpu_usage INTEGER,  -- Total CPU usage
+  disk_usage INTEGER, -- Total disk usage
+  PRIMARY KEY (time)
+);
+
+SELECT * FROM create_hypertable('devices', 'time',
+                                chunk_time_interval => 1000);
+```
+
+To define a continuous aggregate on a table that is using an integer
+dimension, it is necessary to have a function to get the current
+timestamp in whatever representation that you are using and set it for
+the hypertable using
+[`set_integer_now_func`][api-set-integer-now-func]. The function can
+be defined as a normal Postgres function, but need to be
+[`STABLE`][pg-func-stable], take no arguments, and should return an
+integer of the same type as the time column in the table. In our case,
+this should suffice:
+
+```sql
+CREATE OR REPLACE FUNCTION current_microfortnight() RETURNS BIGINT
+LANGUAGE SQL STABLE AS $$
+	SELECT CAST(1209600 * EXTRACT(EPOCH FROM CURRENT_TIMESTAMP) / 1000000 AS BIGINT)
+$$;
+
+SELECT set_integer_now_func('devices', 'current_microfortnight');
+```
+
+Once the replacement for current timestamp has been set up, you can
+define the continuous aggregate for the `devices` table.
+
+```sql
+CREATE VIEW devices_summary
+WITH (timescaledb.continuous,
+      timescaledb.refresh_interval = '1 minute',
+      timescaledb.refresh_lag = '-500') AS
```

Comment:
Unclear why we are using this `refresh_lag` setting. Doesn't this make the aggregate very inefficient (refresh always)?

Just worried that users are following this example because they don't understand these settings (I am myself confused by them) and then don't get the performance they expect.

It is also confusing that the `refresh_interval` and `refresh_lag` are using different time representations.

Reply:
If we do not set the `refresh_lag` the view will not be updated, which might give the appearance that the view is not working. A refresh lag set to this value just means that the view is updated for recent values, so AFAICT that would mean that the "most recent" value might not be accurate, but not affect performance.

Than the two parameters use different time representations does not seem strange to me. The refresh interval gives how often the refresh is executed and should be given in wall time, but the `refresh_lag` is referring to microfortnights in this example, so it makes sense that it is an integer.

Reply:
In other places in our docs, we state there are caveats with setting the refresh_lag like this and it is not the default setting. For instance, our docs says: `Tuning the refresh_lag parameter lower will mean that the aggregates will follow inserts more closely, but can cause some write amplification which may degrade insert performance. `. And, further, the example setting here is not only a "magic number" but also negative!, which I think is confusing for the reader. So, in this example we're basically not only ignoring our own best practices, but we're also confusing the user with very non-intuitive settings.

With default settings, isn't it only the most recent bucket that is not updated? Can't we handle that by explaining in the tutorial that this is how continuous aggs work?

I think it is better to explain the recommended behavior than trying to circumvent it ourselves.